PlutoSDR_Streaming.cpp

```diff
@@ -460,65 +470,87 @@ int tx_streamer::send(	const void * const *buffs,
 
 {
 	std::lock_guard<std::mutex> lock(mutex);
-	size_t items = std::min(4096, (int)numElems);
+	size_t items = std::min(buf_size - items_in_buf, numElems);
 
-	if (buffer.size() != items) {
-		buffer.reserve(items);
-		buffer.resize(items);
-	}
+	int16_t src = 0;
+	uintptr_t dst_ptr, src_ptr = (uintptr_t)&src;
+	ptrdiff_t buf_step = iio_buffer_step(buf);
 
 	for (unsigned int i = 0; i < channel_list.size(); i++) {
+		iio_channel *chn = channel_list[i];
 		unsigned int index = i / 2;
-		if(format==SOAPY_SDR_CS16){
+
+		dst_ptr = (uintptr_t)iio_buffer_first(buf, chn) + items_in_buf * buf_step;
+
+		if (format == SOAPY_SDR_CS16) {
 
 			int16_t *samples_cs16 = (int16_t *)buffs[index];
 			for (size_t j = 0; j < items; ++j) {
-				buffer[j]=samples_cs16[j*2+i];
+				src = samples_cs16[j*2+i];
+				iio_channel_convert_inverse(chn, (void *)dst_ptr, (const void *)src_ptr);
+				dst_ptr += buf_step;
 			}
-
-			channel_write(channel_list[i],buffer.data(), items * sizeof(int16_t));
-
-		}else if(format==SOAPY_SDR_CF32){
+		}
+		else if (format == SOAPY_SDR_CF32) {
 
 			float *samples_cf32 = (float *)buffs[index];
 			for (size_t j = 0; j < items; ++j) {
-				buffer[j]=(int16_t)(samples_cf32[j*2+i]*2048);
+				src = (int16_t)(samples_cf32[j*2+i]*2048);
+				iio_channel_convert_inverse(chn, (void *)dst_ptr, (const void *)src_ptr);
+				dst_ptr += buf_step;
 			}
-			channel_write(channel_list[i],buffer.data(), items * sizeof(int16_t));
-
 		}
 		else if (format == SOAPY_SDR_CS8) {
 
 			int8_t *samples_cs8 = (int8_t *)buffs[index];
 			for (size_t j = 0; j < items; ++j) {
-				buffer[j] = (int16_t)(samples_cs8[j*2 + i] <<4);
+				src = (int16_t)(samples_cs8[j*2 + i] <<4);
+				iio_channel_convert_inverse(chn, (void *)dst_ptr, (const void *)src_ptr);
+				dst_ptr += buf_step;
 			}
-			channel_write(channel_list[i], buffer.data(), items * sizeof(int16_t));
 		}
-
 	}
 
-	ssize_t ret = iio_buffer_push(buf);
+	items_in_buf += items;
 
-	if (ret < 0){
+	if (items_in_buf == buf_size || flags & SOAPY_SDR_END_BURST) {
```

Comment:
Because items may be less than numElems, we only want to flush when we really consumed the entire buffer provided by the user.

if (items_in_buf == buf_size || (flags & SOAPY_SDR_END_BURST && numElems == items)) {